adapters/tlsio_appleios.c

```diff
@@ -0,0 +1,546 @@
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_049: [ If the tlsio_handle is NULL, the tlsio_appleios_close shall not do anything, and return _LINE_. ]*/
+        LogError("NULL TLS handle.");
+        result = __FAILURE__;
+    }
+    else
+    {
+        /* Codes_SRS_TLSIO_APPLEIOS_32_045: [ The tlsio_appleios_close shall store the provided on_io_close_complete callback function address. ]*/
+        tlsio_instance->on_io_close_complete = on_io_close_complete;
+        /* Codes_SRS_TLSIO_APPLEIOS_32_046: [ The tlsio_appleios_close shall store the provided on_io_close_complete_context handle. ]*/
+        tlsio_instance->on_io_close_complete_context = on_io_close_complete_context;
+
+        if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSED) || (tlsio_instance->state == TLSIO_APPLEIOS_STATE_ERROR))
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_079: [ If the tlsio state is TLSIO_APPLEIOS_STATE_ERROR, or TLSIO_APPLEIOS_STATE_CLOSED, the tlsio_appleios_close shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSE, and return 0. ]*/
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_CLOSED;
+            result = 0;
+        }
+        else if ((tlsio_instance->state == TLSIO_APPLEIOS_STATE_OPENING) || (tlsio_instance->state == TLSIO_APPLEIOS_STATE_CLOSING))
+        {
+            /* Codes_SRS_TLSIO_APPLEIOS_32_048: [ If the tlsio state is TLSIO_APPLEIOS_STATE_OPENING, or TLSIO_APPLEIOS_STATE_CLOSING, the tlsio_appleios_close shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSE, and return _LINE_. ]*/
+            LogError("Try to close the connection with an already closed TLS.");
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_ERROR;
+            result = __FAILURE__;
+        }
+        else
+        {
+            CFReadStreamClose(tlsio_instance->sockRead);
+            CFWriteStreamClose(tlsio_instance->sockWrite);
+            /* Codes_SRS_TLSIO_APPLEIOS_32_047: [ If tlsio_appleios_close get success to start the process to close the ssl connection, it shall set the tlsio state as TLSIO_APPLEIOS_STATE_CLOSING, and return 0. ]*/
+            tlsio_instance->state = TLSIO_APPLEIOS_STATE_CLOSING;
+            result = 0;
+            /* Codes_SRS_TLSIO_APPLEIOS_32_044: [ The tlsio_appleios_close shall set the tlsio to try to close the connection for 10 times before assuming that close connection failed. ]*/
+            tlsio_instance->countTry = MAX_TLS_CLOSING_RETRY;
+            /* Codes_SRS_TLSIO_APPLEIOS_32_050: [ If the tlsio_appleios_close get success closing the tls connection, it shall call the tlsio_appleios_dowork. ]*/
+            tlsio_appleios_dowork(tlsio_handle);
+        }
+    }
+
+    return result;
+}
+
+/* Codes_SRS_TLSIO_APPLEIOS_32_052: [ The tlsio_appleios_send shall send all bytes in a buffer to the ssl connectio. ]*/
+int tlsio_appleios_send(CONCRETE_IO_HANDLE tlsio_handle, const void* buffer, size_t size, ON_SEND_COMPLETE on_send_complete, void* callback_context)
```

Comment:
>send [](start = 19, length = 4)

I think it would be nice to call this send_async. Same applies to open. It should be open_async. And so on. better name and it does not require the change of the xio interface.